model/riscv_insts_vext_vset.sail

```diff
@@ -0,0 +1,202 @@
+
+/* ******************************************************************************* */
+/* This file implements part of the vector extension.                              */
+/* Chapter 6: Configuration-Setting Instructions                                   */
+/* ******************************************************************************* */
+
+mapping sew_flag : string <-> bits(3) = {
+  "e8"    <-> 0b000,
+  "e16"   <-> 0b001,
+  "e32"   <-> 0b010,
+  "e64"   <-> 0b011,
+  "e128"  <-> 0b100,
+  "e256"  <-> 0b101,
+  "e512"  <-> 0b110,
+  "e1024" <-> 0b111
+} 
+
+mapping maybe_lmul_flag : string <-> bits(3) = {
+  ""              <-> 0b000, /* m1 by default */
+  sep() ^ "mf8"   <-> 0b101,
+  sep() ^ "mf4"   <-> 0b110,
+  sep() ^ "mf2"   <-> 0b111,
+  sep() ^ "m1"    <-> 0b000,
+  sep() ^ "m2"    <-> 0b001,
+  sep() ^ "m4"    <-> 0b010,
+  sep() ^ "m8"    <-> 0b011
+}
+
+mapping maybe_ta_flag : string <-> bits(1) = {
+  ""           <-> 0b0, /* tu by default */
+  sep() ^ "ta" <-> 0b1,
+  sep() ^ "tu" <-> 0b0
+}
+
+mapping maybe_ma_flag : string <-> bits(1) = {
+  ""           <-> 0b0, /* mu by default */
+  sep() ^ "ma" <-> 0b1,
+  sep() ^ "mu" <-> 0b0
+}
+
+/* ****************************** vsetvli & vsetvl ******************************* */
+union clause ast = VSET_TYPE : (vsetop, bits(1), bits(1), bits(3), bits(3), regidx, regidx)
+
+mapping encdec_vsetop : vsetop <-> bits(4) ={
+  VSETVLI <-> 0b0000,
+  VSETVL  <-> 0b1000
+}
+
+mapping clause encdec = VSET_TYPE(op, ma, ta, sew, lmul, rs1, rd) if haveRVV() 
+  <-> encdec_vsetop(op) @ ma @ ta @ sew @ lmul @ rs1 @ 0b111 @ rd @ 0b1010111 if haveRVV()
+
+function clause execute VSET_TYPE(op, ma, ta, sew, lmul, rs1, rd) = {
+  let VLEN_pow      = get_vlen_pow();
+  let ELEN_pow      = get_elen_pow();
+  let LMUL_pow_ori  = get_lmul_pow();
+  let SEW_pow_ori   = get_sew_pow(); 
+  let ratio_pow_ori = SEW_pow_ori - LMUL_pow_ori;
+
+  /* set vtype */
+  match op {
+    VSETVLI => {
+      vtype->bits() = 0b0 @ zeros(sizeof(xlen) - 9) @ ma @ ta @ sew @ lmul
+    },
+    VSETVL  => {
+      let rs2 : regidx = sew[1 .. 0] @ lmul;
+      vtype->bits() = X(rs2)
+    }
+  };
+
+  /* check legal SEW and LMUL and calculate VLMAX */
+  let LMUL_pow_new = get_lmul_pow();
+  let SEW_pow_new  = get_sew_pow();
+  if SEW_pow_new > LMUL_pow_new + ELEN_pow then {
+    vtype->bits() = 0b1 @ zeros(sizeof(xlen) - 1); /* set vtype.vill */
+    vl = EXTZ(0b0);
+    print_reg("CSR vtype <- " ^ BitStr(vtype.bits()));
+    print_reg("CSR vl <- " ^ BitStr(vl));
+    return RETIRE_SUCCESS
+  };
+  let VLMAX = int_power(2, VLEN_pow + LMUL_pow_new - SEW_pow_new);
+
+  /* set vl according to VLMAX and AVL */
+  if (rs1 != 0b00000) then { /* normal stripmining */
+    let rs1_val = X(rs1);
+    let AVL = unsigned(rs1_val);
+    vl = if AVL <= VLMAX then to_bits(sizeof(xlen), AVL)
+         else if AVL < 2 * VLMAX then to_bits(sizeof(xlen), (AVL + 1) / 2) /* ceil(AVL / 2) ≤ vl ≤ VLMAX */
```

Comment:
This is a place where the specification is too open and makes very difficult configurability. Some implementations may choose to use VLMAX rather than ceil(AVL / 2). Probably no implementation will use any other value, but the specification allows them. I guess this matches spike. Perhaps should add a note for a future configurability option.

Reply:
I have added a note in riscv_insts_vext_vset.sail for this issue.